docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md

```diff
@@ -0,0 +1,121 @@
+
+## Decision Drivers
+
+* We want to get changes merged and released to our consumer applications faster.
+* We want to simplify pushing new code to applications, especially urgent bug fixes.
+* We want to move towards the future monorepo world.
+* Mozilla is moving towards rapid releases for mobile code.
+* We are willing to accept the overhead of upgrading application-services more frequently in application repositories.
+
+## Considered Options
+
+* **(A) Keep our current release process**
+* **(B) Move to a rapid release process where changes in app-services are frequently merged into consumer applications**
+    * Phase 1: update our branching model and release process (Jan 2023 - Feb 2023)
+        * Update our branches to match how firefox-android works: a `main` branch that is used for the current nightly builds and a `releases_v[N]` for each previous release.
+        * Automate a nightly release of our `main` branch
+          * See the discussion at the bottom of this document for how this could work
+          * Note: regardless of which option we choose, we will no longer be using semver-style versions and also not following semver semantics.  We will be releasing breaking changes in a nightly cycle without bumping the major version.
+        * Manually handle patch releases from the `release_v*` branches
+          * Update our version number to `[firefox-version].[N+1]`
+          * Create a GitHub release from the branch
+          * We should avoid having unreleased code in these branches.  Releases should be been soon after new code is committed to those branches.
+          * Code should always be uplifted into these branches.  Before code is merged into the beta branch it should have already been merged into `main` and released as a nightly.  Before code is merged into `release_v[N-1]` branch, it should have merged and released in the `release_v[N]` branch.
+        * Continue to maintain our CHANGELOG, but with new versioning.  In main, keep a rolling log of updates with the header `[firefox-version].0`
+        * This replaces our current release process.
+        * To push changes to consumers:
+           * Create a PR in their repository that bumps the application-services version to one of these releases.
+           * Don't merge any of these PRs unless all pass CI and are approved.
```

Comment:
Slightly tangential to Ryan's point -- what is the benefit of waiting for, eg: the PR in firefox-ios to be approved before merging an update for fenix? I can certainly see the benefit of waiting for CI to pass in all repos, but I'm having trouble understanding why we'd want to wait on review everywhere.

Reply:
What happens now is that we get one PR approved and merged, say on Android, but then the iOS PR can sit around for a long period of time before being approved.  The reasons vary: maybe it's a large change on the iOS side, maybe they want to think through changes that affect the UX, maybe they just don't have the bandwith to review something.

In any case, what ends up happening that Android and iOS have very different versions of app-services, which can get confusing.  Also one of them is blocked from taking any new code until the PR is reviewed.